### PR TITLE
Regex, objects, bits

### PR DESCRIPTION
**Pull request by anderseknert**
Signed-off-by: Anders Eknert <anders@eknert.com>

### CODE REVIEW

File: src/test/clojure/jarl/builtins/regex_test.clj [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+(deftest builtin-regex-split-test
+  (testing "regex.split"
+    (is (= (builtin-regex-split "[\\d]" "a1b2c") ["a" "b" "c"]))))
+
+(deftest builtin-regex-find-n-test
+  (testing "regex.find_n"
+    (is (= (builtin-regex-find-n "[\\d]" "a1b2c3" 3) ["1" "2" "3"]))
+    (is (= (builtin-regex-find-n "[\\d]" "a1b2c3" -1) ["1" "2" "3"]))
+    (is (= (builtin-regex-find-n "[\\d]" "a1b2c3" 2) ["1" "2"]))
```
Comment (johanfylling) [inline src/test/clojure/jarl/builtins/regex_test.clj:23]: `0` as the `number` arg could be an additional interesting edge case to test.

File: src/main/clojure/jarl/builtins/regex.clj [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  (let [matcher (.matcher (Pattern/compile pattern) value)]
+    (loop [i 0
+           matches []
+           match (.find matcher)]
+      (if (or (false? match) (and (not= number -1) (>= i number)))
+        matches
+        (recur (inc i) (conj matches (.group matcher 0)) (.find matcher))))))
+
+; Annoyingly, this method is available in the Google package, but not made public :/
+; https://github.com/google/re2j/blob/master/java/com/google/re2j/RE2.java#L883
```
Comment (johanfylling) [inline src/main/clojure/jarl/builtins/regex.clj:41]: Bummer. Should be possible to find a suitable solution. A task for another day 😄 .

File: src/test/clojure/jarl/builtins/strings_test.clj [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+                                           builtin-indexof builtin-indexof-n builtin-lower builtin-replace
+                                           builtin-strings-reverse builtin-split builtin-startswith builtin-substring
+                                           builtin-trim builtin-trim-left builtin-trim-prefix builtin-trim-right
+                                           builtin-trim-suffix builtin-trim-space builtin-upper]])
   (:import (se.fylling.jarl BuiltinException)))
 
+(deftest builtin-concat-test
+  (testing "concat"
+    (is (= (builtin-concat ", " ["a" "b" "c"]) "a, b, c"))
+    (is (= (builtin-concat "🙂" ["🙃" "🙃" "🙃"]) "🙃🙂🙃🙂🙃"))))
```
Comment (johanfylling) [inline src/test/clojure/jarl/builtins/strings_test.clj:13]: 😆

**Review by johanfylling — APPROVED** [review]
A nice set of additions to the ever growing number of supported built-ins 😃 .

File: src/main/clojure/jarl/builtins/bits.clj [hunk 4]
```diff
@@ -0,0 +1,72 @@
+(ns jarl.builtins.bits
+  (:require [jarl.builtins.utils :refer [undefined-ex java->rego]]))
+
+(defn- non-integer? [num]
+  (not (zero? (mod num 1))))
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/bits.clj:5]: Indeed. I think this might even have been done in the other PR, so holding off on this until that's been merged.

File: src/test/clojure/jarl/builtins/bits_test.clj [hunk 5]
```diff
@@ -0,0 +1,96 @@
+(ns jarl.builtins.bits_test
+  (:require [clojure.test :refer [deftest is testing]]
+            [jarl.builtins.bits :refer [builtin-bits-or builtin-bits-and builtin-bits-negate
+                                        builtin-bits-xor builtin-bits-lsh builtin-bits-rsh]])
+  (:import (se.fylling.jarl UndefinedException)))
+
+(deftest builtin-bits-or-test
+  (testing "bits.or"
+    (is (= (builtin-bits-or 13 10) 15))
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/test/clojure/jarl/builtins/bits_test.clj:9]: Yes, but it would not mirror what we'd see in OPA, would it? :)

File: src/main/clojure/jarl/builtins/bits.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/bits.clj:5]: Indeed, I've stopped adding type-checks to the builtins here as I think we'll want to have that in the evaluator instead, based on the argument types provided by the plan for each builtin. WDYT?

File: src/main/clojure/jarl/builtins/objects.clj [hunk 6]
```diff
@@ -0,0 +1,58 @@
+(ns jarl.builtins.objects
+  (:require [clojure.string :as str]
+            [jarl.builtins.utils :refer [str->int]]))
+
+(defn builtin-object-get
+  "Implementation of object.get built-in"
+  {:builtin "object.get" :args-types ["object", "any", "any"]}
+  [object key default]
+  (if (vector? key)
+    (get-in object key default)
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/objects.clj:10]: indeed :D

File: src/main/clojure/jarl/builtins/regex.clj (hunk 2 above)
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/regex.clj:41]: For sure! I'm still not sure why they'd do that!

File: src/test/clojure/jarl/builtins/bits_test.clj [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    (is (= (builtin-bits-or 10 -10) -2))
+    (is (= (builtin-bits-or -10 -10) -10)))
+  (testing "non-integers"
+    (is (thrown-with-msg? UndefinedException
+                          #"eval_type_error: bits.or: operand 1 must be integer number but got floating-point number"
+                          (builtin-bits-or 10.1 1)))
+    (is (thrown-with-msg? UndefinedException
+                          #"eval_type_error: bits.or: operand 2 must be integer number but got floating-point number"
+                          (builtin-bits-or 2 2.22)))))
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/test/clojure/jarl/builtins/bits_test.clj:18]: Yeah my current thinking is that we'll remove this from each individual built-in in favor for a more general purpose check in the evaluator when calling _any_ built-in. The built-ins may still provide more granularity though, just like in OPA.. so if the evaluator checks for "number" we may still have the built-in check if it's an integer, and so on.

File: src/test/clojure/jarl/builtins/objects_test.clj [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  (testing "object.union_n"
+    (is (= (builtin-object-union-n [{"foo" "bar"} {"foo" "baz"}]) {"foo" "baz"}))
+    (is (= (builtin-object-union-n [{"foo" "bar"} {"foo" "bar" "x" 1}]) {"foo" "bar" "x" 1}))
+    (is (= (builtin-object-union-n [{"foo" "bar"} {"foo" "baz"} {"x" 1} {"y" 2 "x" 10}]) {"foo" "baz" "x" 10 "y" 2}))))
+
+(deftest builtin-object-filter-test
+  (testing "object.filter"
+    (is (= (builtin-object-filter {"a" 1 "b" 2 "c" 3} ["b" "c"]) {"b" 2 "c" 3}))
+    (is (= (builtin-object-filter {"a" 1 "b" 2 "c" 3} {"b" 2}) {"b" 2}))
+    (is (= (builtin-object-filter {"a" 1 "b" 2 "c" 3} {}) {}))))
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply src/test/clojure/jarl/builtins/objects_test.clj:35]: Yep, `json.remove` does this.

File: src/test/clojure/jarl/builtins/regex_test.clj (hunk 1 above)
Comment (anderseknert) [reply src/test/clojure/jarl/builtins/regex_test.clj:24]: Sure! Added.
